### ISSUE TITLE
why F.softmax(tensor.float()) for onnx_trace while F.softmax(tensor,dtype=torch.float()) if onnx_trace is false?

### ISSUE THREAD

**Issue by songtaoshi** [pytorch/fairseq#3518]
#### What is your question?
I am confused why when onnx_trace we should using **x.float()** rather than **F.softmax(...,dtype=torch.floadt32)**
https://github.com/pytorch/fairseq/blob/b0ae834d528a4a466202107a22356aed71bb6161/fairseq/utils.py#L487

#### Code
```python
def softmax(x, dim: int, onnx_trace: bool = False):
    if onnx_trace:
        return F.softmax(x.float(), dim=dim)
    else:
        return F.softmax(x, dim=dim, dtype=torch.float32)
```

**Comment by songtaoshi** [comment]
@ynd @noe @zpao @gazay

**Comment by lematt1991** [comment]
CC @myleott, looks like 48d9afbeb37584586411c3d8e53d91228528dec4 is where it got added

**Comment by songtaoshi** [comment]
@lematt1991 @myleott how to understand the code ?

**Comment by shabie** [comment]
Fair question. No update?
